packages/chain-controller/src/SnapHandlerClient.ts

```diff
@@ -0,0 +1,130 @@
+import type { JsonRpcRequest } from '@metamask/keyring-api/dist/JsonRpcRequest';
+import type { SnapController } from '@metamask/snaps-controllers';
+import type { SnapId } from '@metamask/snaps-sdk';
+import { HandlerType } from '@metamask/snaps-utils';
+import type { Json } from '@metamask/utils';
+import { v4 as uuid } from 'uuid';
+
+/**
+ * Handler for Snap requests.
+ */
+export type Handler = SnapController['handleRequest'];
+
+/**
+ * Send requests to a Snap through a Snap request handler.
+ */
+class SnapHandlerSender {
+  #snapId: SnapId;
+
+  #origin: string;
+
+  #handler: Handler;
+
+  #handlerType: HandlerType;
+
+  /**
+   * Constructor for `SnapHandlerSender`.
+   *
+   * @param handler - The Snap request handler to send requests to.
+   * @param handlerType - The handler type.
+   * @param snapId - The ID of the snap to use.
+   * @param origin - The sender's origin.
+   */
+  constructor(
+    handler: Handler,
+    handlerType: HandlerType,
+    snapId: SnapId,
+    origin: string,
+  ) {
+    this.#snapId = snapId;
+    this.#origin = origin;
+    this.#handler = handler;
+    this.#handlerType = handlerType;
+  }
+
+  /**
+   * Sends a request to the snap and return the response.
+   *
+   * @param request - JSON-RPC request to send to the snap.
+   * @returns A promise that resolves to the response of the request.
+   */
+  async send(request: JsonRpcRequest): Promise<unknown> {
+    return this.#handler({
+      snapId: this.#snapId,
+      origin: this.#origin,
+      handler: this.#handlerType,
+      request,
+    });
+  }
+}
+
+/**
+ * Snap client to submit requests through a handler that submit requests to
+ * a Snap.
+ */
+export class SnapHandlerClient {
+  #handler: Handler;
+
+  #sender: SnapHandlerSender;
+
+  /**
+   * Constructor for SnapHandlerClient.
+   *
+   * @param options - The client options.
+   * @param options.handler - A function to submit requests to the Snap handler
+   * (this should call the SnapController.handleRequest)
+   * @param options.snapId - The Snap ID.
+   * @param options.origin - The origin from which the Snap is being invoked.
+   */
+  constructor({
+    handler,
+    // Follow same pattern than for @metamask/keyring-api
+    snapId = 'undefined' as SnapId,
```

Comment:
What is the purpose of this?

Reply:
This is also coming from the `keyring-api`, the rationale being that you can construct a client without providing any `snapId` and you can then constraint it using `withSnapId` method.

Also, if we were to use an optional here, that would require us to "over-check" if `.snapId` is defined or not, which will result in a more complex code logic IMO.

Anyway, if you submit a request with an invalid Snap id, the request will fail, so we're cover here (not sure if that's the best pattern though...).

As a side note, I just re-used the same pattern than for the `keyring-api` too 😅

Reply:
That being said, after re-reading the logic, I guess I could get rid of this pattern...

Lemme try to address that, cause I feel like this is gonna make things a bit simpler and straightforward!

Reply:
> _if we were to use an optional here, that would require us to "over-check" if .snapId_

I think I don't follow what you mean here. Is this just to avoid the TS error because of an optional property? Because setting it to `undefined` will be the same as having it as optional in terms of execution